mne/beamformer/_dics.py

```diff
@@ -1,27 +1,581 @@
+    if normalize_fwd:
+        if inversion == 'single':
+            fwd_norm = 'dipole'
+        else:
+            fwd_norm = 'vertex'
+    else:
+        fwd_norm = None  # No normalization
+
+    picks = _setup_picks(info=info, forward=forward)
+    _, ch_names, proj, vertices, G = _prepare_beamformer_input(
+        info, forward, label, picks=picks, pick_ori=pick_ori,
+        fwd_norm=fwd_norm,
+    )
+    csd_picks = [csd.ch_names.index(ch) for ch in ch_names]
+
+    n_sources = G.shape[1] // n_orient
+
+    logger.info('Computing DICS spatial filters...')
+    Ws = []
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    computing DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_data(index=i)
+
+        if real_filter:
+            Cm = Cm.real
+
+        # Ensure the CSD is in the same order as the leadfield
+        Cm = Cm[csd_picks, :][:, csd_picks]
+
+        # Tikhonov regularization using reg parameter to control for
+        # trade-off between spatial resolution and noise sensitivity
+        # eq. 25 in Gross and Ioannides, 1999 Phys. Med. Biol. 44 2081
+        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+
+        if (pick_ori == 'max-power' and weight_norm == 'unit-noise-gain' and
+                inversion == 'matrix'):
+            Cm_inv_sq = Cm_inv.dot(Cm_inv)
+
+        # Compute spatial filters
+        W = np.dot(G.T, Cm_inv)
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            Gk = G[:, n_orient * k: n_orient * k + n_orient]
+
+            # Compute power at the source
+            Ck = np.dot(Wk, Gk)
+
+            # Normalize the spatial filters
+            if Wk.ndim == 2 and len(Wk) > 1:
+                # Free source orientation
+                if inversion == 'single':
+                    # Invert for each dipole separately using plain division
+                    Wk /= np.diag(Ck)[:, np.newaxis]
+                elif inversion == 'matrix':
+                    # Invert for all dipoles simultaneously using matrix
+                    # inversion.
+                    Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
+            else:
+                # Fixed source orientation
+                Wk /= Ck
+
+            if pick_ori == 'max-power':
+                # Compute spectral power, so we can pick the orientation that
+                # maximizes this power.
+                if inversion == 'matrix' and weight_norm == 'unit-noise-gain':
+                    # Use Eq. 4.47 from [2]_
+                    norm_inv = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                    if reduce_rank:
+                        # Use pseudo inverse computation setting smallest
+                        # component to zero if the leadfield is not full rank
+                        norm = _eig_inv(norm_inv, norm_inv.shape[0] - 1)
```

Comment:
This also originates from `make_lcmv`. @britta-wstnr and @agramfort were doing simulations on phantom data and ran into an inversion issue here. I don't know exactly why they choose to solve it in this manner instead of using `rcond`, so maybe they can pitch in on this.

Reply:
We ran into a problem with complex eigenvalues within the computation of the optimal orientation when using ``pinv``  if the leadfield was only rank 2 (e.g., with the spherical headmodel of the phantom data), compare https://github.com/mne-tools/mne-python/pull/4568 and https://github.com/mne-tools/mne-python/pull/4628. For LCMV, this approach works well and has the advantage of not having to find a reasonable tolerance value. I don't know, however, to what extent it is directly translatable to DICS.

Reply:
Yes it looks like I had the same comment in #4568 but nobody answered there, either :)

@wmvanvliet okay if I push directly a few `# XXX` reminder comments about these things?

Reply:
> okay if I push directly a few # XXX reminder comments about these things?

go ahead!